File: services/webapp/app/types/workzone.ts
// Type definitions for Workzone API

// Main Workzone entity
export interface Workzone {
  id: string;
  org_id: string;
  name: string;
  description?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

// Payload for creating a workzone
export interface WorkzoneCreate {
  org_id: string;
  name: string;
  description?: string;
  metadata?: Record<string, any>;
}

// Payload for updating a workzone
export interface WorkzoneUpdate {
  name?: string;
  description?: string;
  metadata?: Record<string, any>;
}

// Response from API for a single workzone
export interface WorkzoneResponse extends Workzone {
  invocation_count?: number;
}

// Response for listing workzones
export interface WorkzoneList {
  workzones: WorkzoneResponse[];
  has_more: boolean;
  total?: number;
}

// Workzone Invocation entity - matches actual API response
export interface WorkzoneInvocation {
  invocation_id: string;
  user_task_id: string;
  assistant_task_id?: string;  // May be null if still processing
  created_at: string;
  workzone_id?: string;
  org_id: string;
  project_id: string;
  agent_id: string;
  status?: 'pending' | 'processing' | 'completed' | 'failed';
  error?: string;
  updated_at?: string;
  metadata?: Record<string, any>;
  duration_ms?: number;
  token_usage?: {
    input_tokens?: number;
    output_tokens?: number;
    total_tokens?: number;
  };
}

// Payload for creating a workzone invocation
export interface InvocationCreate {
  org_id: string;
  project_id: string;
  agent_id: string;
  user_id: string;
  content: string;
  metadata?: Record<string, any>;
}

// Response from API for a single invocation
export interface InvocationResponse extends WorkzoneInvocation {
  cost_usd?: number;
}

// List item for invocation listing
export interface InvocationListItem {
  invocation_id: string;
  user_task_id: string;
  assistant_task_id: string;
  org_id: string;
  project_id: string;
  agent_id: string;
  created_at: string;
}

// Response for listing invocations
export interface InvocationList {
  invocations: InvocationListItem[];
  has_more: boolean;
  total?: number;
}

// Type guards for runtime type checking
export function isWorkzone(obj: any): obj is Workzone {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    typeof obj.id === 'string' &&
    typeof obj.org_id === 'string' &&
    typeof obj.name === 'string'
  );
}

export function isWorkzoneResponse(obj: any): obj is WorkzoneResponse {
  return isWorkzone(obj);
}

export function isWorkzoneList(obj: any): obj is WorkzoneList {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    Array.isArray(obj.workzones) &&
    obj.workzones.every(isWorkzoneResponse) &&
    typeof obj.has_more === 'boolean'
  );
}

export function isWorkzoneInvocation(obj: any): obj is WorkzoneInvocation {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    typeof obj.invocation_id === 'string' &&
    typeof obj.user_task_id === 'string' &&
    typeof obj.created_at === 'string'
  );
}

export function isInvocationResponse(obj: any): obj is InvocationResponse {
  return isWorkzoneInvocation(obj);
}

export function isInvocationList(obj: any): obj is InvocationList {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    Array.isArray(obj.invocations) &&
    obj.invocations.every(isWorkzoneInvocation) &&
    typeof obj.has_more === 'boolean'
  );
}

// ============================================
// Streaming message types for workzone
// ============================================

export interface ThinkingMessage {
  type: 'thinking';
  thinking: string;
  signature?: string;
  timestamp: number;
}

export interface TextMessage {
  type: 'text';
  text: string;
  timestamp: number;
}

export interface ToolExecutionMessage {
  type: 'tool';
  toolId: string;
  toolName: string;
  input: Record<string, any>;
  result?: any;
  isError?: boolean;
  timestamp: number;
}

export type StreamingMessage = ThinkingMessage | TextMessage | ToolExecutionMessage;

File: services/backend-api/app/models/workzone.py
from typing import Optional, Dict, Any
from datetime import datetime
import uuid

from beanie import Document
from pydantic import BaseModel, Field

from app.models.base import BaseModel as ChicoryBaseModel


class Workzone(Document):
    """Workzone document model"""
    id: str = Field(default_factory=lambda: str(uuid.uuid4()), alias="_id")
    org_id: str
    name: str
    description: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "workzones"

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True


class WorkzoneInvocation(Document):
    """Workzone Invocation document model"""
    id: str = Field(default_factory=lambda: str(uuid.uuid4()), alias="_id")
    workzone_id: str
    project_id: str
    agent_id: str
    user_id: str
    user_task_id: str
    assistant_task_id: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "workzone_invocations"

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True


# Request/Response Models
class WorkzoneCreate(BaseModel):
    org_id: str = Field(..., min_length=1)
    name: str
    description: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class WorkzoneUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class WorkzoneResponse(BaseModel):
    id: str
    org_id: str
    name: str
    description: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime


class WorkzoneList(BaseModel):
    workzones: list[WorkzoneResponse]
    has_more: bool


class InvocationCreate(BaseModel):
    org_id: str
    project_id: str
    agent_id: str
    user_id: str
    content: str
    metadata: Optional[Dict[str, Any]] = None


class InvocationResponse(BaseModel):
    invocation_id: str
    user_task_id: str
    assistant_task_id: str


class InvocationListItem(BaseModel):
    invocation_id: str
    user_task_id: str
    assistant_task_id: str
    org_id: str
    project_id: str
    agent_id: str
    user_id: str
    created_at: datetime


class InvocationList(BaseModel):
    invocations: list[InvocationListItem]
    has_more: bool
